### PR TITLE
Processing Algorithm "Open file or URL"

### PR DESCRIPTION
**Pull request by signedav**
A simple native algorithm to be able to open local files in the corresponding programs or URLs in the web browser.

image

Useful in combination with the Download File Algorithm to open the resulting file or displaying results after models triggering external systems.

### CODE REVIEW

File: src/analysis/processing/qgsalgorithmurlopener.cpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+  bool result = QDesktopServices::openUrl( qurl );
+
+  if ( result )
+    feedback->pushInfo( QObject::tr( "Successfully opened %1" ).arg( url ) );
+  else
+    feedback->pushInfo( QObject::tr( "Failed opening %1" ).arg( url ) );
+
+  QVariantMap outputs;
+  outputs.insert( QStringLiteral( "OUTPUT" ), result );
```
Comment (Gustry) [inline src/analysis/processing/qgsalgorithmurlopener.cpp:82]: Is-it missing the QgsProcessingOutputBoolean in the initAlgorithm ?

Comment (signedav) [reply src/analysis/processing/qgsalgorithmurlopener.cpp:82]: not it's there. You are right QgsProcessingOutputBoolean makes sense here.  I'll change that.

File: src/analysis/processing/qgsalgorithmurlopener.cpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+#include "qgis.h"
+
+#include <QUrl>
+#include <QDesktopServices>
+
+///@cond PRIVATE
+
+QString QgsUrlOpenerAlgorithm::name() const
+{
+  return QStringLiteral( "urlopener" );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmurlopener.cpp:29]:
```suggestion
  return QStringLiteral( "openurl" );
```

And rename the class to "QgsOpenUrlAlgorithm". "URL Opener" is a bit too FME-like for me, and doesn't follow the naming convention of our other algorithms. (Eg we use "DXF export", not "DXF exporter", and "Filter by geometry type" not "Geometry type filterer".)

File: src/analysis/processing/qgsalgorithmurlopener.cpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+QString QgsUrlOpenerAlgorithm::groupId() const
+{
+  return QStringLiteral( "filetools" );
+}
+
+QString QgsUrlOpenerAlgorithm::shortHelpString() const
+{
+  return tr( "This algorithm opens files in the corresponding programs or URLs in the webbrowser" );
+}
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmurlopener.cpp:55]: Can you add shortDescription too?

File: src/analysis/processing/qgsalgorithmurlopener.cpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+QString QgsUrlOpenerAlgorithm::groupId() const
+{
+  return QStringLiteral( "filetools" );
+}
+
+QString QgsUrlOpenerAlgorithm::shortHelpString() const
+{
+  return tr( "This algorithm opens files in the corresponding programs or URLs in the webbrowser" );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmurlopener.cpp:54]:
```suggestion
  return tr( "This algorithm opens files in their default associated application, or URLs in the user's default web browser." );
```

File: src/analysis/processing/qgsalgorithmurlopener.cpp [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+QgsUrlOpenerAlgorithm *QgsUrlOpenerAlgorithm::createInstance() const
+{
+  return new QgsUrlOpenerAlgorithm();
+}
+
+void QgsUrlOpenerAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterString( QStringLiteral( "URL" ), tr( "URL or file path" ), QStringLiteral( "https://www.qgis.org/" ), false, false ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmurlopener.cpp:64]: I don't think we should have any default here:

```suggestion
  addParameter( new QgsProcessingParameterString( QStringLiteral( "URL" ), tr( "URL or file path" ), QVariant(), false, false ) );
```

File: src/analysis/processing/qgsalgorithmurlopener.cpp [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+QVariantMap QgsUrlOpenerAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  QString url = parameterAsString( parameters, QStringLiteral( "URL" ), context );
+  if ( url.isEmpty() )
+    throw QgsProcessingException( tr( "No URL or file path specified" ) );
+  QUrl qurl = QUrl::fromUserInput( url );
+
+  bool result = QDesktopServices::openUrl( qurl );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmurlopener.cpp:75]:
```suggestion
  const QString url = parameterAsString( parameters, QStringLiteral( "URL" ), context );
  if ( url.isEmpty() )
    throw QgsProcessingException( tr( "No URL or file path specified" ) );
  const QUrl qurl = QUrl::fromUserInput( url );

  const bool result = QDesktopServices::openUrl( qurl );
```

File: src/analysis/processing/qgsalgorithmurlopener.cpp [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+  if ( result )
+    feedback->pushInfo( QObject::tr( "Successfully opened %1" ).arg( url ) );
+  else
+    feedback->pushInfo( QObject::tr( "Failed opening %1" ).arg( url ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmurlopener.cpp:80]:
```suggestion
    feedback->reportError( QObject::tr( "Failed opening %1" ).arg( url ) );
```

**Comment by nyalldawson** [comment]
Nice feature! :+1:

**Comment by nirvn** [comment]
Nice one @signedav

**Comment by signedav** [comment]
Thanks  @nyalldawson I agree on all inputs and concerned.

File: src/analysis/processing/qgsalgorithmurlopener.cpp [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+QgsOpenUrlAlgorithm *QgsOpenUrlAlgorithm::createInstance() const
+{
+  return new QgsOpenUrlAlgorithm();
+}
+
+void QgsOpenUrlAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterString( QStringLiteral( "URL" ), tr( "URL or file path" ), QVariant(), false, false ) );
+  addOutput( new QgsProcessingOutputBoolean( QStringLiteral( "SUCCESS" ), QObject::tr( "Successfully performed opening file or url" ) ) );
```
Comment (Gustry) [inline src/analysis/processing/qgsalgorithmurlopener.cpp:70]:
```suggestion
  addOutput( new QgsProcessingOutputBoolean( QStringLiteral( "SUCCESS" ), QObject::tr( "Successfully performed opening file or URL" ) ) );
```

**Review by Gustry — APPROVED** [review]
Thanks, nice addition.
Just minor fix.

**Pull request merged**